activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -94,7 +94,11 @@ def self.[](*args)
     #
     # This value can be later fetched using either +:key+ or +'key'+.
     def []=(key, value)
-      regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      if value.frozen?
+        regular_writer(convert_key(key), convert_value(value))
+      else
+        regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      end
```

Comment:
:for        | frozen | duped 
-------------|--------|-------
 :assignment | true   | true  
 :assignment | false  | false 
 :to_hash    | true   | true  
 :to_hash    | false  | true  
 nil         | true   | true  
 nil         | false  | true  

It's 8am in the morning and I've only had the one cup of coffee so my logic may be fuzzy :smile: